glue-codes/openfast-cpp/src/FAST_Prog.cpp

```diff
@@ -11,23 +11,61 @@ inline bool checkFileExists(const std::string& name) {
 void readTurbineData(int iTurb, fast::fastInputs & fi, YAML::Node turbNode) {
 
   //Read turbine data for a given turbine using the YAML node
-  fi.globTurbineData[iTurb].TurbID = turbNode["turb_id"].as<int>();
-  fi.globTurbineData[iTurb].FASTInputFileName = turbNode["FAST_input_filename"].as<std::string>() ;
-  fi.globTurbineData[iTurb].FASTRestartFileName = turbNode["restart_filename"].as<std::string>() ;
+  if (turbNode["turb_id"]) {
+      fi.globTurbineData[iTurb].TurbID = turbNode["turb_id"].as<int>();
+  } else {
+      turbNode["turb_id"] = iTurb;
+  }
+  if (turbNode["FAST_input_filename"]) {
+      fi.globTurbineData[iTurb].FASTInputFileName = turbNode["FAST_input_filename"].as<std::string>() ;  
+  } else {
+      fi.globTurbineData[iTurb].FASTInputFileName = "";
+  }
+  if (turbNode["restart_filename"]) {
+      fi.globTurbineData[iTurb].FASTRestartFileName = turbNode["restart_filename"].as<std::string>() ;
+  } else {
+      fi.globTurbineData[iTurb].FASTRestartFileName = "";
+  }
+  if ( (fi.globTurbineData[iTurb].FASTRestartFileName == "") && (fi.globTurbineData[iTurb].FASTInputFileName == "") )
+      throw std::runtime_error("Both FAST_input_filename and restart_filename are empty or not specified for Turbine " + std::to_string(iTurb));
   if (turbNode["turbine_base_pos"].IsSequence() ) {
-    fi.globTurbineData[iTurb].TurbineBasePos = turbNode["turbine_base_pos"].as<std::vector<double> >() ;
+      fi.globTurbineData[iTurb].TurbineBasePos = turbNode["turbine_base_pos"].as<std::vector<float> >() ;
+  } else {
+      fi.globTurbineData[iTurb].TurbineBasePos = std::vector<float>(3,0.0);
   }
   if (turbNode["turbine_hub_pos"].IsSequence() ) {
-    fi.globTurbineData[iTurb].TurbineHubPos = turbNode["turbine_hub_pos"].as<std::vector<double> >() ;
+      fi.globTurbineData[iTurb].TurbineHubPos = turbNode["turbine_hub_pos"].as<std::vector<double> >() ;
+  } else {
+      fi.globTurbineData[iTurb].TurbineHubPos =  std::vector<double>(3,0.0);
+  }
+  if (turbNode["num_force_pts_blade"]) {
```

Comment:
It might be better to throw error on this (i.e., make these options mandatory) instead of defaulting to values that are not correct.

Reply:
Agreed. I rectified a few of these automatic assumptions in 9f34f9eb119d38dc0c78bfe58fdd23d084d02430. However, a user shouldn't have to specify the turbine location for running a standard OpenFAST simulation through the C++ API with `InflowType` as `1`  in the OpenFAST `.fst` input file (specified using the InflowWind module). For actuator line simulations `InflowType` must be `2` and would require a turbine location to be specified. Unfortunately, I wouldn't know if a user specified  or `2` until the OpenFAST is initialized.